pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -305,81 +305,64 @@ func (a Adapter) waitAndGetComponentPod(hideSpinner bool) (*corev1.Pod, error) {
-	if len(pushDevfileCommands) == 0 {
+	if len(commandsMap) == 0 {
 		return errors.New(fmt.Sprint("error executing devfile commands - there should be at least 1 command"))
 	}
 
-	commandOrder := []common.CommandNames{}
+	compInfo := common.ComponentInfo{
+		PodName: podName,
+	}
 
 	// Only add runinit to the expected commands if the component doesn't already exist
 	// This would be the case when first running the container
 	if !componentExists {
-		commandOrder = append(commandOrder, common.CommandNames{DefaultName: string(common.DefaultDevfileInitCommand), AdapterName: a.devfileInitCmd})
-	}
-	commandOrder = append(
-		commandOrder,
-		common.CommandNames{DefaultName: string(common.DefaultDevfileBuildCommand), AdapterName: a.devfileBuildCmd},
-		common.CommandNames{DefaultName: string(common.DefaultDevfileRunCommand), AdapterName: a.devfileRunCmd},
-	)
-
-	// Loop through each of the expected commands in the devfile
-	for i, currentCommand := range commandOrder {
-		// Loop through each of the command given from the devfile
-		for _, command := range pushDevfileCommands {
-			// If the current command from the devfile is the currently expected command from the devfile
-			if command.Name == currentCommand.DefaultName || command.Name == currentCommand.AdapterName {
-				// If the current command is not the last command in the slice
-				// it is not expected to be the run command
-				if i < len(commandOrder)-1 {
-					// Any exec command such as "Init" and "Build"
-
-					for _, action := range command.Actions {
-						compInfo := common.ComponentInfo{
-							ContainerName: *action.Component,
-							PodName:       podName,
-						}
-
-						err = exec.ExecuteDevfileBuildAction(&a.Client, action, command.Name, compInfo, show)
-						if err != nil {
-							return err
-						}
-					}
-
-					// If the current command is the last command in the slice
-					// it is expected to be the run command
-				} else {
-					// Last command is "Run"
-					klog.V(4).Infof("Executing devfile command %v", command.Name)
-
-					for _, action := range command.Actions {
-
-						// Check if the devfile run component containers have supervisord as the entrypoint.
-						// Start the supervisord if the odo component does not exist
-						if !componentExists {
-							err = a.InitRunContainerSupervisord(*action.Component, podName, containers)
-							if err != nil {
-								return
-							}
-						}
-
-						compInfo := common.ComponentInfo{
-							ContainerName: *action.Component,
-							PodName:       podName,
-						}
-
-						if componentExists && !common.IsRestartRequired(command) {
-							klog.V(4).Infof("restart:false, Not restarting DevRun Command")
-							err = exec.ExecuteDevfileRunActionWithoutRestart(&a.Client, action, command.Name, compInfo, show)
-							return
-						}
-
-						err = exec.ExecuteDevfileRunAction(&a.Client, action, command.Name, compInfo, show)
-					}
-				}
+		// Get Init Command
```

Comment:
I have a few questions here with which you might help

1. Do we want to have three separate if blocks and execute the commands in order; or do we want to loop the map and use the previous idea of having a predefined order and executing it?
2. What can we do about refactoring code between Kubernetes Adapter and Docker Adapter execDevfile() since they're pretty much the same. See my comment here https://github.com/openshift/odo/issues/3015#issuecomment-634071947

Reply:
> I have a few questions here with which you might help
> 
> 1. Do we want to have three separate if blocks and execute the commands in order; or do we want to loop the map and use the previous idea of having a predefined order and execute the commands?

Predefined order is not required as we have `kind` associated with each command, for now lets leave it like this, It has to be updated with `events` support.

> 2. What can we do about refactoring code between Kubernetes Adapter and Docker Adapter execDevfile() since they're pretty much the same. See my comment here [#3015 (comment)](https://github.com/openshift/odo/issues/3015#issuecomment-634071947)

Agreed, that can be refactored.